libcudacxx/include/cuda/std/__bit/countr.h

```diff
@@ -21,91 +21,65 @@
 #  pragma system_header
 #endif // no system header
 
+#include <cuda/std/__bit/bit_cast.h>
 #include <cuda/std/__bit/ctz.h>
-#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__concepts/concept_macros.h>
+#include <cuda/std/__type_traits/conditional.h>
+#include <cuda/std/__type_traits/is_constant_evaluated.h>
 #include <cuda/std/__type_traits/is_unsigned_integer.h>
 #include <cuda/std/cstdint>
 #include <cuda/std/limits>
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-// Forward decl for recursive use in split word operations
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr int __countr_zero(_Tp __t) noexcept;
-
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<sizeof(_Tp) <= sizeof(uint32_t), int>
-__countr_zero_dispatch(_Tp __t) noexcept
-{
-  return __cccl_ctz(static_cast<uint32_t>(__t));
-}
-
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<sizeof(_Tp) == sizeof(uint64_t), int>
-__countr_zero_dispatch(_Tp __t) noexcept
-{
-  return __cccl_ctz(static_cast<uint64_t>(__t));
-}
-
-template <typename _Tp, int _St = sizeof(_Tp) / sizeof(uint64_t)>
-struct __countr_zero_rsh_impl
-{
-  static _LIBCUDACXX_HIDE_FROM_ABI constexpr int __short_circuit(_Tp __t, int __cur, int __count)
-  {
-    // Stops processing early if non-zero
-    return (__cur == numeric_limits<uint64_t>::digits)
-           ? __countr_zero_rsh_impl<_Tp, _St - 1>::__count(__t, __cur + __count)
-           : __cur + __count;
-  }
-
-  static _LIBCUDACXX_HIDE_FROM_ABI constexpr int __count(_Tp __t, int __count)
-  {
-    return __short_circuit(__t >> numeric_limits<uint64_t>::digits, __countr_zero(static_cast<uint64_t>(__t)), __count);
-  }
-};
-
-template <typename _Tp>
-struct __countr_zero_rsh_impl<_Tp, 1>
+_CCCL_TEMPLATE(class _Tp)
+_CCCL_REQUIRES(_CCCL_TRAIT(_CUDA_VSTD::__cccl_is_unsigned_integer, _Tp) _CCCL_AND(sizeof(_Tp) <= sizeof(uint64_t)))
+_LIBCUDACXX_HIDE_FROM_ABI constexpr int __countr_zero(_Tp __t) noexcept
 {
-  static _LIBCUDACXX_HIDE_FROM_ABI constexpr int __count(_Tp __t, int __count)
+  if (is_constant_evaluated() && __t == 0)
   {
-    return __count + __countr_zero(static_cast<uint64_t>(__t));
+    return numeric_limits<_Tp>::digits;
   }
-};
-
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<(sizeof(_Tp) > sizeof(uint64_t)), int>
-__countr_zero_dispatch(_Tp __t) noexcept
-{
-  return __countr_zero_rsh_impl<_Tp>::__count(__t, 0);
+  using _Sp         = _If<sizeof(_Tp) <= sizeof(uint32_t), uint32_t, uint64_t>;
+  auto __ctz_result = _CUDA_VSTD::__cccl_ctz(static_cast<_Sp>(__t));
+  NV_IF_ELSE_TARGET(NV_IS_DEVICE,
+                    // if __t == 0 __ctz_result is already equal to numeric_limits<_Tp>::digits
+                    (return sizeof(_Tp) < sizeof(uint32_t) && __t == 0 ? numeric_limits<_Tp>::digits : __ctz_result;),
+                    (return __t ? __ctz_result : numeric_limits<_Tp>::digits;))
```

Comment:
All conditionals in this function - and the other ones - check for `__t == 0` whereas this checks for `__t` and flips the order

Please change these to also use the same condition